compiler/rustc_target/src/spec/mod.rs

```diff
@@ -1803,6 +1803,7 @@ supported_targets! {
 
     ("wasm32-unknown-emscripten", wasm32_unknown_emscripten),
     ("wasm32-unknown-unknown", wasm32_unknown_unknown),
+    ("wasm32v1-none", wasm32v1_none),
```

Comment:
Will the way to cfg on this specific target be to specify:
```rust
#[cfg(all(target_family = "wasm", target_os = "none"))]
```
or
```rust
#[cfg(all(target_arch = "wasm32", target_os = "none"))]
```

If in the future a `wasm32v2` is released, how will it be distinguished in a `cfg`?

Or will the `target_arch` be `wasm32v1` and in the future `wasm32v2`?

Reply:
You can just differentiate them through the target features.

Reply:
Could you provide an example?

At the moment differentiating using `target_features` seems unreliable, because it's common for target features to become released but their cfg to be still marked as unstable and is therefore not usable in stable rust, as has been the case with:
- https://github.com/rust-lang/rust/pull/131080

Reply:
FWIW the documentation in this PR [has an answer to your original question](https://github.com/rust-lang/rust/pull/131487/files#diff-cb13cc0f59e5c9559f00dfaa9afec2b266a09ff54a803ab200a6c1162c3b9c0bR33):

> It's recommended to conditionally compile code for this target with:
>
> ```text
> #[cfg(all(target_family = "wasm", target_os = "none"))]
> ```

Otherwise I would also agree with @CryZe. All other native platforms have a constant `target_arch` string with a variable `target_feature` set corresponding to levels of support for various features.

The situation you're referring to where `multivalue` and `reference-types` aren't stable right now isn't an issue for this target. This'll be a new target when it lands and users should also be able to see `multivalue` and `reference-types` when they get to use it since that PR should land relatively soon as well.

Stabilization of wasm features has lagged behind for awhile and this is basically good motivation to ensure we keep up in the future. For example another option is blocking a hypothetical `wasm32v2-none` target until all of the features it enables are landed and stable in Rust. (or, for example, blocking this PR on #131080 but to be clear I'm not advocating we do that)

Reply:
Will the `target_env` be set to `v1` or `v2`?

Reply:
I do not expect us to update `env` based on the architecture version, no.

Reply:
As Alex said: The incentive structure that encourages people to push for stabilizing a target feature name that is well-defined, so that we can e.g. evaluate it for its ABI impacts, is intentional.

Reply:
I think what @leighmcculloch is looking for is a way to make a cfg check that specifically identifies the condition of "being limited to the features in `wasm32v1-none`". Like he wants to not detect any specific new features, but detect the _absence_ of _all_ new features, including features not-yet-invented. Because that's the only case we want to compile for. We want to emit an error if the user tries to build for a target that enables any further wasm features. Which -- even if the current set of feature names all stabilize -- he can't write code to detect because they don't exist yet.

If there were -- speaking hypothetically -- some way to say `cfg(target_triple="wasm32v1-none")` or even `cfg(target_subarch="wasm32v1")` that would do the trick. Or anything else that pins to this subarch version _and not future versions_. I'm not sure if there's a good way to make that happen with the tools currently available though.

Reply:
I believe there is nonesuch in ways to conveniently do that, yes.

The answer, however, is `cfg(target_feature = "nontrapping-fptoint")`.

The reason that will work is that feature was specifically added **for Rust**, *and* it is part of the v2 definition, so "post v1" Rust targets simply will not ever exclude it.

Reply:
Mhm. Thanks. I'm sure we'll figure out some not-so-bad guardrails that handle the majority of cases.